pkg/controllers/federatedhpa/util.go

```diff
@@ -0,0 +1,420 @@
+)
+
+type Resource struct {
+	gvk       schema.GroupVersionKind
+	namespace string
+	name      string
+}
+
+type handleCacheResp struct {
+	hpaFTC            *fedcorev1a1.FederatedTypeConfig
+	hpaGVR            schema.GroupVersionResource
+	hpaObject         *unstructured.Unstructured
+	fedHPAObject      fedcorev1a1.GenericFederatedObject
+	fedWorkloadObject fedcorev1a1.GenericFederatedObject
+	ppObject          fedcorev1a1.GenericPropagationPolicy
+	newPPResource     *Resource
+}
+
+func (r Resource) QualifiedName() common.QualifiedName {
+	return common.QualifiedName{
+		Namespace: r.namespace,
+		Name:      r.name,
+	}
+}
+
+func fedObjectToSourceObjectResource(object metav1.Object) (Resource, error) {
+	fedObj, _ := object.(fedcorev1a1.GenericFederatedObject)
+	metadata, err := fedObj.GetSpec().GetTemplateMetadata()
+	if err != nil {
+		return Resource{}, err
+	}
+	return Resource{
+		name:      metadata.GetName(),
+		namespace: metadata.GetNamespace(),
+		gvk:       metadata.GroupVersionKind(),
+	}, nil
+}
+
+func policyObjectToResource(object metav1.Object) Resource {
+	if cpp, ok := object.(*fedcorev1a1.PropagationPolicy); ok {
+		return Resource{
+			name:      cpp.GetName(),
+			namespace: cpp.GetNamespace(),
+			gvk: schema.GroupVersionKind{
+				Group:   fedcorev1a1.SchemeGroupVersion.Group,
+				Version: fedcorev1a1.SchemeGroupVersion.Version,
+				Kind:    PropagationPolicyKind,
+			},
+		}
+	}
+
+	if pp, ok := object.(*fedcorev1a1.ClusterPropagationPolicy); ok {
+		return Resource{
+			name:      pp.GetName(),
+			namespace: pp.GetNamespace(),
+			gvk: schema.GroupVersionKind{
+				Group:   fedcorev1a1.SchemeGroupVersion.Group,
+				Version: fedcorev1a1.SchemeGroupVersion.Version,
+				Kind:    ClusterPropagationPolicyKind,
+			},
+		}
+	}
+
+	return Resource{}
+}
+
+func generateFederationHPANotWorkReason(newPPResource *Resource, pp fedcorev1a1.GenericPropagationPolicy) string {
+	var reasons []string
+	if newPPResource == nil {
+		reasons = append(reasons, "The workload is not bound to any propagationPolicy.")
+		return fmt.Sprintf("%v", reasons)
+	}
+
+	ppKind, ppName := newPPResource.gvk.Kind, newPPResource.name
+	if !isPropagationPolicyExist(pp) {
+		reasons = append(reasons, fmt.Sprintf("The %s %s bound to the workload does not exist.", ppKind, ppName))
+		return fmt.Sprintf("%v", reasons)
+	}
+	if isPropagationPolicyExist(pp) && !isPropagationPolicyDividedMode(pp) {
+		reasons = append(reasons, fmt.Sprintf("The %s %s bound to the workload is not Divided mode.", ppKind, ppName))
+	}
+
+	return fmt.Sprintf("%v", reasons)
+}
+
+func generateDistributedHPANotWorkReason(
+	ctx context.Context,
+	newPPResource *Resource,
+	pp fedcorev1a1.GenericPropagationPolicy,
+	fedWorkload fedcorev1a1.GenericFederatedObject,
+	hpaObject *unstructured.Unstructured,
+) string {
+	var reasons []string
+
+	if !isWorkloadRetainReplicas(fedWorkload) {
+		reasons = append(reasons, "The workload is not enabled for retain replicas.")
```

Comment:
since we have `fedWorkload` here, why don't we print GVK and name of the workload in reasons?

Reply:
User can found it from ScaleTargetRef. So it seems unnecessary.